runtime/permissions/mod.rs

```diff
@@ -34,6 +36,8 @@ pub use prompter::PromptCallback;
 static DEBUG_LOG_ENABLED: Lazy<bool> =
   Lazy::new(|| log::log_enabled!(log::Level::Debug));
 
+pub type PermissionsContainer = Arc<Mutex<Permissions>>;
```

Comment:
If every usage of Permissions is replaced with PermissionsContainer, then why not just include that in the actual struct. IE
```
struct Permissions(Arc<Mutex<Inner>>)
```

Reply:
Not every usage is replaced with `PermissionsContainer` - usages in `ext/` crates are using `Arc<Mutex<P>>` which depend on traits from various crates. I tried the approach you suggest but it doesn't worked.

Reply:
> usages in ext/ crates are using Arc<Mutex<P>> which depend on traits from various crates

what's an example?

Reply:
Examples:
- `ext/net/ops.rs`
- `ext/net/ops_tls.rs`
- `ext/fetch/lib.rs`

These ops don't have access to `Permissions` struct (because `runtime/` crate depends on `ext/net` and `ext/fetch`) and thus to `PermissionsContainer`. Extension crates define traits for permission checks that have matching signatures that are later implemented in `runtime/` for `Permissions` struct (and thus `PermissionsContainer` since it derefs to `Permissions`). This is somewhat optimal solution because it doesn't introduce tight coupling between `ext/` crates and `runtime/` thanks to Rust's trait system.

Reply:
I don't get it. Let's take `op_net_send_udp` for example. Here it gets a NetPermissions object from the state:

https://github.com/denoland/deno/blob/319f6074761421b797db71bf10f6171516e3d92a/ext/net/ops.rs#L166-L169

How is changing Permissions to be `struct Permissions(Arc<Mutex<Inner>>)` going to affect that?

Reply:
I guess that might also work - in such case in your suggestion the traits would be implemented on outer `Permissions` struct that would take care of locking. Only downside I see to it is that in some cases we would need to acquire mutex lock multiple times for a single ops.

Reply:
Yeah, let’s make this a struct that buries this type (could make PermissionsContainer a struct). Lots of noise.

Reply:
Rewrote it

Reply:
There's still a PermissionsContainer tho?

Reply:
To reduce scenarios where we acquire a lock multiple times, I believe we still need one type for the unlocked state then one for the locked state.

One thing we could maybe do is hide the mutex within the unlocked state type's interface, but I'm not sure it's worth it.

Reply:
> There's still a PermissionsContainer tho?

Yes, there are still situations where you need to get to the inner structure, eg. in CLI - I'd much rather handle that via `Permissions` struct than some `Inner`.